README.md
# Lego Mindstorms EV3

[![CircleCI](https://circleci.com/gh/nerves-project/nerves_system_ev3.svg?style=svg)](https://circleci.com/gh/nerves-project/nerves_system_ev3)
[![Hex version](https://img.shields.io/hexpm/v/nerves_system_ev3.svg "Hex version")](https://hex.pm/packages/nerves_system_ev3)

This is the base Nerves System configuration for the Lego Mindstorms EV3 brick.

![EV3 brick image](assets/images/lego-mindstorms-ev3.jpg)
<br><sup>[Image credit](#wikipediaref)</sup>

| Feature              | Description                     |
| -------------------- | ------------------------------- |
| CPU                  | 300 MHz ARM926EJ-S              |
| Memory               | 64 MB DRAM                      |
| Storage              | 16 MB Flash and MicroSD         |
| Linux kernel         | 4.4 w/ ev3dev patches           |
| Display              | 178x128 LCD - /dev/fb0          |
| IEx terminal         | USB Gadget port or port 1 (see below) |
| GPIO, I2C, SPI       | Yes - ev3dev drivers            |
| ADC                  | Yes                             |
| PWM                  | Yes, but no Elixir support      |
| UART                 | 4 available - ttyACM0           |
| Speakers             | Built-in speaker - ALSA         |
| Camera               | None                            |
| Ethernet             | Requires USB Ethernet dongle    |
| WiFi                 | Requires USB WiFi dongle        |
| Bluetooth            | Not supported                   |

## Using

The most common way of using this Nerves System is create a project with `mix
nerves.new` and to export `MIX_TARGET=ev3`. See the [Getting started
guide](https://hexdocs.pm/nerves/getting-started.html#creating-a-new-nerves-app)
for more information.

If you need custom modifications to this system for your device, clone this
repository and update as described in [Making custom
systems](https://hexdocs.pm/nerves/systems.html#customizing-your-own-nerves-system)

I recommend creating a project that uses
[nerves_init_gadget](https://github.com/nerves-project/nerves_init_gadget) to
get started on the EV3. This makes using the EV3 similar to the Raspberry Pi
Zero so more help will be available on the Nerves slack channel and forum.

Ethernet for uploading new firmware images is provided through the USB gadget
port (labelled "PC" on the EV3). If you connect this via a USB cable to your
laptop and turn on the EV3, you should eventually see a new USB and Ethernet
port. (The first boot of Nerves on the EV3 takes some time.)

Unfortunately, the console doesn't appear to work via the Gadget USB port. I
haven't had time to debug this, so I left it going through `ttyS1` which comes
out port 1. This is very sad. If you have a clue on how to fix this, see
`rootfs_overlay/etc/erlinit.config` to route the console through `ttyGS0` and
let me know.

The Lego device kernel modules are not built into the kernel so they need to be
loaded by your application at initialization time. To do this, run the following
manually or add them to your application:

```elixir
iex> :os.cmd('modprobe suart_emu')
iex> :os.cmd('modprobe legoev3_ports')
iex> :os.cmd('modprobe snd_legoev3')
iex> :os.cmd('modprobe legoev3_battery')
```

When Nerves supports Bluetooth, you'll want to run the following line as well:

```elixir
iex> :os.cmd('modprobe legoev3_bluetooth')
```

Once you get to the console and can update firmware, you should be able to
follow ev3dev instructions for reading and writing to files to control motors,
read sensors, and everything else that the EV3 can do.

## Example projects

Since the documentation is sparse here, you may find some example projects
helpful:

* [nerves_ev3_example](https://github.com/fhunleth/nerves_ev3_example)

If you have a project to share, please help us by adding it to the list and
sending a pull request. Thanks!

## Port 1 console access

The EV3 supports a special UART output on port 1. Nerves can use this output for
Linux kernel debug messages and an IEx prompt. If you plan on doing any Linux
kernel, driver, or boot related work with the EV3, this console is
indispensable.

You will either need to buy a [console
adapter](http://www.mindsensors.com/ev3-and-nxt/40-console-adapter-for-ev3) or
build [one](http://botbench.com/blog/2013/08/15/ev3-creating-console-cable/) to
use this port.

Loading the `legoev3_ports` driver automatically disables the console port.
Since we're working on the EV3 and it's not as easy to use as it should be,
we've told the `legoev3_ports` module to not touch it. If you're on the EV3,
you'll see the following line in `/etc/modprobe.d/ev3dev.conf`:

    options legoev3_ports disable_in_port=1

If you want to use port 1, you'll need to disable this. To do this, add a
`/etc/modprobe.d/ev3dev.conf` to your project's `rootfs_overlay`. If an empty
file exists, it will override this default one, but I usually create a file with
the line commented out so that I remember what the special line is.

## Supported USB WiFi devices

The base image includes drivers and firmware for Ralink RT53xx (`rt2800usb`
driver), MediaTek MT7601U (`mt7601u`), and Edimax EW-7811Un (`8192cu`) devices.
One option for these devices is to get a Tenda W311MI Wireless USB Adapter.

We are still working out which subset of all possible WiFi dongles to support in
our images. At some point, we may have the option to support all dongles and
selectively install modules at packaging time, but until then, these drivers and
their associated firmware blobs add significantly to Nerves release images.

If you are unsure what driver your WiFi dongle requires, run Raspbian and
configure WiFi for your device. At a shell prompt, run `lsmod` to see which
drivers are loaded.  Running `dmesg` may also give a clue. When using `dmesg`,
reinsert the USB dongle to generate new log messages if you don't see them.

## Wired Ethernet

If you have a USB Ethernet adapter, find the driver for it on your PC. For
example, plug it in and check `dmesg` and `lsmod` to see which driver it loads.
In my case, I have an adapter that loads the `asix` driver. Make sure that your
driver is compiled in as a module to the Linux kernel in Nerves and then
manually load the driver via `modprobe asix`.

## Power down, halt, reboot

On most platforms, the default behavior of Nerves is to hang if something goes
wrong.  This is good for debugging since rebooting makes it easy to lose console
messages or it might hide the issue completely. On the EV3 hanging requires a
slightly more complex restarting process.

1. Hold down the Back, and center buttons on the EV3 Brick
1. When the screen goes blank, release the Back button
1. When the screen says “Starting,” release the center button

If you would like to power down instead you will need to change `erlinit.config`
from `--hang-on-exit` to `---poweroff-on-exit`.

If you're attached to the console, you may see a kernel panic when you run power
off. From what I can tell, this panic happens after the important parts of
shutting down gracefully have completed and does not cause a problem.

## SDCard vs. internal NAND Flash notes

The EV3 brick has a 16 MB NAND Flash inside it that's connected to SPI bus 0.
It doesn't look like the ev3dev project has included support for it yet except
in their version of u-boot. The means that it can only be programmed using the
Lego supplied tools. The 16 MB NAND Flash also has a couple other issues. First,
it appears to be super slow. This leads to them copying the whole image to DRAM
instead of reading it as needed. It appears that this uses up 10 MB of DRAM
compared to running off the SDCard. This is significant when you consider that
the board only has 64 MB total DRAM. On the other hand, programming the internal
NAND Flash is cool and the direction that we'd prefer to go on production
systems.

Currently, the u-boot in the internal NAND Flash that's supplied by Lego and the
ev3dev project expects the `uImage` in the first VFAT partition. Ideally, it
would extract it out of the rootfs so that we could implement more atomic
firmware updates. To avoid the need to reflash the firmware to use Nerves, I'm
staying with the existing mechanism.

## ev3dev

This port draws substantially on the [ev3dev](http://www.ev3dev.org/) project.
In general, if there's a way to do something in ev3dev, it can be made to work
in Nerves. Nerves uses the same Linux kernel from the ev3dev project and enables
the same set of custom drivers that were created by the ev3dev developers.

## Provisioning devices

This system supports storing provisioning information in a small key-value store
outside of any filesystem. Provisioning is an optional step and reasonable
defaults are provided if this is missing.

Provisioning information can be queried using the Nerves.Runtime KV store's
[`Nerves.Runtime.KV.get/1`](https://hexdocs.pm/nerves_runtime/Nerves.Runtime.KV.html#get/1)
function.

Keys used by this system are:

Key                    | Example Value     | Description
:--------------------- | :---------------- | :----------
`nerves_serial_number` | "1234578"`        | By default, this string is used to create unique hostnames and Erlang node names. If unset, it defaults to part of the EV3's device ID.

The normal procedure would be to set these keys once in manufacturing or before
deployment and then leave them alone.

For example, to provision a serial number on a running device, run the following
and reboot:

```elixir
iex> cmd("fw_setenv nerves_serial_number 1234")
```

This system supports setting the serial number offline. To do this, set the
`NERVES_SERIAL_NUMBER` environment variable when burning the firmware. If you're
programming MicroSD cards using `fwup`, the commandline is:

```sh
sudo NERVES_SERIAL_NUMBER=1234 fwup path_to_firmware.fw
```

Serial numbers are stored on the MicroSD card so if the MicroSD card is
replaced, the serial number will need to be reprogrammed. The numbers are stored
in a U-boot environment block. This is a special region that is separate from
the application partition so reformatting the application partition will not
lose the serial number or any other data stored in this block.

Additional key value pairs can be provisioned by overriding the default provisioning.conf
file location by setting the environment variable 
`NERVES_PROVISIONING=/path/to/provisioning.conf`. The default provisioning.conf
will set the `nerves_serial_number`, if you override the location to this file,
you will be responsible for setting this yourself.

[Image credit](#wikipediaref): By Klaus-Dieter Keller - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=29156877


CHANGELOG.md
# Changelog

## v1.4.0

This release contains various updates to provisioning variables and data.

**Host requirements**

Building firmware using this system requires `fwup` to be updated on your
host computer to at least `v1.2.5`. The target minimum version requirement 
has not changed from `0.15.0`.

**Serial numbers**

Device serial numbers are now set using `NERVES_SERIAL_NUMBER` instead of 
`SERIAL_NUMBER`. This is to reduce ambiguity on the source of the serial
by name spacing it along side other Nerves variables. The U-Boot environment
key has also changed from `serial_number` to `nerves_serial_number`. The 
erlinit.config has been updated to provide backwards compatibility for setting
the hostname from the serial number by checking for `nerves_serial_number`
and falling back to `serial_number`.

**Custom provisioning**

Provisioning data is applied at the time of calling the `fwup` task `complete`.
The `complete` task is executed when writing the firmware to the target disk.
During this time, `fwup` will include the contents of a provisioning file
located at `${NERVES_SYSTEM}/images/fwup_include/provisioning.conf`. By default,
this file only sets `nerves_serial_number`. You can add additional provisioning
data by overriding the location of this file to include your own by setting
the environment variable `NERVES_PROVISIONING`. If you override this variable
you will be responsible for also setting `nerves_serial_number`.

* Updated dependencies
  * [nerves_system_br v1.4.5](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.4.5)

## v1.3.0

This release upgrades gcc from version 6.3.0 to 7.3.0. See the toolchain release
notes for more information.

* Bug fixes
  * Disable `ldattach` and build package `dumb_ldattach`.
    This fixes an issue with that prevents the UART sensor from functioning.

* Updated dependencies
  * [nerves_system_br v1.4.1](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.4.1)
  * [nerves_toolchain_armv5tejl_unknown_linux_musleabi v1.1.0](https://github.com/nerves-project/toolchains/releases/tag/v1.1.0)

## v1.2.1

* Updated dependencies
  * [nerves_system_br v1.3.2](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.3.2)

## v1.2.0

This release updates Erlang to OTP 21.0

* Updated dependencies
  * [nerves_system_br v1.3.0](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.3.0)

## v1.1.1

This release fixes some issues and adds firmware UUID support. This support can
be used to unambiguously know what's running on a device.

* Updated dependencies
  * [nerves_system_br v1.2.2](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.2.0)

* Bug fixes
  * Empty serial numbers stored in the U-Boot environment would be used instead
    of reverting to devices IDs built into the CPU or board.
  * It wasn't possible to enable QtWebEngine (needed for kiosk apps, but don't do this on an EV3)

## v1.1.0

This release adds official support for provisioning serial numbers to devices.
Other information can be provisioned in a similar manner. See the README.md for
details.

Buildroot was also updated to 2018.05. Be sure to review the `nerves_system_br`
link for the changes in the embedded Linux components.

* Updated dependencies
  * [nerves_system_br v1.2.0](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.2.0)

## v1.0.0

This release is nearly identical to rc.1 except for pulling the official 1.0
versions of Nerves dependencies and minor documentation updates.

* Updated dependencies
  * [nerves_system_br v1.0.0](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.0.0)
  * [nerves_toolchain v1.0.0](https://github.com/nerves-project/toolchains/releases/tag/v1.0.0)
  * [nerves v1.0.0](https://github.com/nerves-project/nerves/releases/tag/1.0.0)

## v1.0.0-rc.1

This release contains updates to Erlang and heart from `nerves_system_br` and
mostly cosmetic changes to this project. The trivial `.fw` files are no longer
created by CI scripts. If you've forked this project and are building systems
using CI, make sure to update your publish scripts.

* Updated dependencies
  * [nerves_system_br v1.0.0-rc.2](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.0.0-rc.2)

## v1.0.0-rc.0

* Updated dependencies
  * [nerves_system_br v1.0.0-rc.0](https://github.com/nerves-project/nerves_system_br/releases/tag/v1.0.0-rc.0)
  * [nerves_toolchain v1.0.0-rc.0](https://github.com/nerves-project/toolchains/releases/tag/v1.0.0-rc.0)
  * [nerves v1.0.0-rc.0](https://github.com/nerves-project/nerves/releases/tag/1.0.0-rc.0)

## v0.15.0

* Updated dependencies
  * [nerves_system_br v0.17.0](https://github.com/nerves-project/nerves_system_br/releases/tag/v0.17.0)
  * [nerves_toolchain v0.13.0](https://github.com/nerves-project/toolchains/releases/tag/v0.13.0)
  * [nerves v0.9.0](https://github.com/nerves-project/nerves/releases/tag/v0.9.0)

## v0.14.2

* Updated dependencies
  * [nerves_system_br v0.16.3](https://github.com/nerves-project/nerves_system_br/releases/tag/v0.16.3)
    This fixes the call to otp_build so that it always uses Buildroot's version
    the autoconf tools.

## v0.14.1

* Updated dependencies
  * [nerves_system_br v0.16.2](https://github.com/nerves-project/nerves_system_br/releases/tag/v0.16.2)

## v0.14.0

This version enables the USB Gadget port (MiniUSB port labelled PC). Firmware
updates can be sent over it. Unfortunately, the virtual serial port doesn't work,
so the special UART cable on port 1 is still needed to see the console by default.

* Updated dependencies
  * [nerves_system_br v0.16.1-2017-11](https://github.com/nerves-project/nerves_system_br/releases/tag/v0.16.1-2017-11)

* Bug fixes
  * Fixed symlink issue that prevented the system from being used from hex.pm

* Enhancements
  * Reboot automatically if Erlang VM exits - This is consistent with other
    Nerves systems. See rootfs_overlay/etc/erlinit.config if undesired.

test/mix.exs
if Mix.env() == :test do
  hash =
    :os.cmd('git rev-parse HEAD')
    |> to_string
    |> String.trim()

  System.put_env("NERVES_FW_VCS_IDENTIFIER", hash)
end

defmodule Test.MixProject do
  use Mix.Project

  def project do
    [
      app: :test,
      version: "0.1.0",
      elixir: "~> 1.4",
      archives: [nerves_bootstrap: "~> 1.0"],
      start_permanent: Mix.env() == :prod,
      aliases: [loadconfig: [&bootstrap/1]],
      deps: deps()
    ]
  end

  # Type `mix help compile.app` to learn about applications.
  def application, do: []

  defp bootstrap(args) do
    System.put_env("MIX_TARGET", "ev3")
    Application.start(:nerves_bootstrap)
    Mix.Task.run("loadconfig", args)
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:nerves_system_ev3, path: "../", runtime: false},
      {:nerves_system_test, github: "nerves-project/nerves_system_test"}
    ]
  end
end


test/config/config.exs
use Mix.Config

# Repository specific configuration

system = :nerves_system_ev3
platform = "ev3"
arch = "arm"

app_part_devparth = "/dev/mmcblk0p4"
network_interface = System.get_env("NERVES_NETWORK_INTERFACE") || "wlan0"

# Environment specific configuration
#  Nerves Project test farm configuration
#  NERVES_TEST_SERVER = nerves-test-server.herokuapp.com
#  WEBSOCKET_PROTOCOL = wss

test_server = System.get_env("NERVES_TEST_SERVER")
websocket_protocol = System.get_env("WEBSOCKET_PROTOCOL") || "ws"

# Common configuration

# Configure shoehorn boot order.
config :shoehorn,
  app: :nerves_system_test,
  init: [:nerves_runtime, :system_registry_term_storage, :nerves_network]

# Only trust signed firmware
config :nerves_system_test, :firmware, public_key: System.get_env("NERVES_FW_PUB_KEY")

# Configure system_registry term storage to store the wifi credentials on the
#  app data partition. If the device is using eth0 as the primary connection
#  mechanism the wlan0 settings do not need to be configured.
config :system_registry, SystemRegistry.TermStorage,
  path: "/root/system_registry",
  scopes: [
    [:config, :network_interface, "wlan0", :ssid],
    [:config, :network_interface, "wlan0", :psk]
  ]

# Configure the default interface settings.
# wlan0 | eth0 - Used to establish a connection to the test server.
# usb0 - configured with linklocal to be validated as part of the test results.
config :nerves_network, :default,
  eth0: [
    ipv4_address_method: :dhcp
  ],
  wlan0: [
    ipv4_address_method: :dhcp
  ],
  usb0: [
    ipv4_address_method: :linklocal
  ]

# Configure the url for the connection to the test server phoenix channel socket.
config :nerves_system_test, NervesTestServer.Socket,
  url: "#{websocket_protocol}://#{test_server}/socket/websocket"

# Configure the test suite. nerves_system_test needs to know information such as
#  system - the name of the system repo the tests are being executed on
#  network_interface - the interface that should be used for reporting the
#   results on
#  tests - the locations for the tests to run.
#    Currently, it is only supported that tests are included in the priv_dir
#    of the app and dependencies. The default layout runs tests that are common
#    across devices (:nerves_system_test) and those that are specific to the
#    device (this app)
config :nerves_system_test,
  system: system,
  network_interface: network_interface,
  tests: [
    {:test, :priv_dir, "test"},
    {:nerves_system_test, :priv_dir, "test"}
  ]

# The configuration stored here is duplicated from the project so it can be
#  validated by nerves_system_test because the source is unavailable at runtime.
config :nerves_runtime, :kv,
  nerves_fw_application_part0_devpath: app_part_devparth,
  nerves_fw_application_part0_fstype: "ext4",
  nerves_fw_application_part0_target: "/root",
  nerves_fw_architecture: arch,
  nerves_fw_author: "The Nerves Team",
  nerves_fw_description: Mix.Project.config()[:description],
  nerves_fw_platform: platform,
  nerves_fw_product: Mix.Project.config()[:app],
  nerves_fw_vcs_identifier: System.get_env("NERVES_FW_VCS_IDENTIFIER"),
  nerves_fw_version: Mix.Project.config()[:version]


mix.exs
defmodule NervesSystemEv3.MixProject do
  use Mix.Project

  @app :nerves_system_ev3
  @version Path.join(__DIR__, "VERSION")
           |> File.read!()
           |> String.trim()

  def project do
    [
      app: @app,
      version: @version,
      elixir: "~> 1.4",
      compilers: Mix.compilers() ++ [:nerves_package],
      nerves_package: nerves_package(),
      description: description(),
      package: package(),
      deps: deps(),
      aliases: [loadconfig: [&bootstrap/1], docs: ["docs", &copy_images/1]],
      docs: [extras: ["README.md"], main: "readme"]
    ]
  end

  def application do
    []
  end

  defp bootstrap(args) do
    System.put_env("MIX_TARGET", "ev3")
    Application.start(:nerves_bootstrap)
    Mix.Task.run("loadconfig", args)
  end

  defp nerves_package do
    [
      type: :system,
      artifact_sites: [
        {:github_releases, "nerves-project/#{@app}"}
      ],
      platform: Nerves.System.BR,
      platform_config: [
        defconfig: "nerves_defconfig"
      ],
      checksum: package_files()
    ]
  end

  defp deps do
    [
      {:nerves, "~> 1.0", runtime: false},
      {:nerves_system_br, "1.4.5", runtime: false},
      {:nerves_toolchain_armv5tejl_unknown_linux_musleabi, "1.1.0", runtime: false},
      {:nerves_system_linter, "~> 0.3.0", runtime: false},
      {:ex_doc, "~> 0.18", only: :dev}
    ]
  end

  defp description do
    """
    Nerves System - Lego Mindstorms EV3 http://www.lego.com/en-us/mindstorms/about-ev3
    """
  end

  defp package do
    [
      maintainers: ["Frank Hunleth", "Justin Schneck"],
      files: package_files(),
      licenses: ["Apache 2.0"],
      links: %{"Github" => "https://github.com/nerves-project/#{@app}"}
    ]
  end

  defp package_files do
    [
      "fwup_include",
      "package",
      "rootfs_overlay",
      "CHANGELOG.md",
      "Config.in",
      "external.mk",
      "fwup-revert.conf",
      "fwup.conf",
      "LICENSE",
      "linux-4.4-ev3dev.defconfig",
      "mix.exs",
      "nerves_defconfig",
      "nerves_logo_178x128.pbm",
      "post-build.sh",
      "post-createfs.sh",
      "README.md",
      "VERSION"
    ]
  end

  # Copy the images referenced by docs, since ex_doc doesn't do this.
  defp copy_images(_) do
    File.cp_r("assets", "doc/assets")
  end
end
